Initial Dart Web run configuration commit, currently disabled in the plugin.xml

jwren opened the pull request:
- the new run configuration is disabled in the `plugin.xml`,
- debugging does not yet work

@alexander-doroshko I am happy to wait until everything is farther along, or to create some smaller PRs along the way for history and review.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java:
```diff
@@ -0,0 +1,53 @@
+      @NotNull
+      @Override
+      public String getName() {
+        return "Dart Web"; // compatibility
```
That's the first version of the new run configuration, there's nothing we need to be compatible with :)

jwren replied:
Removed

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/DartBundle.properties:
```diff
@@ -98,10 +98,17 @@ fields.to.generate.setters=Select Fields to Generate Setters
 remote.debug.configuration.description=Debug Remote Dart Command Line Application
+
+webdev.debug.configuration.name=Dart Web
+webdev.debug.configuration.description=Debug Dart Webdev Application
```
AFAIU, webdev is just a tool that helps to deal with Dart web applications. Does 'Webdev Application' make any sense?

jwren replied:
Good point, updated to `Debug Dart Webdev`, I was just trying to be consistent... `Debug using Dart Webdev` would also be fine.

devoncarew replied:
Or just, 'Debug Dart web application'?

alexander-doroshko replied:
>Or just, 'Debug Dart web application'?

👍 That was my thought too.
Or 'Debug Dart web application using the webdev tool'

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java:
```diff
@@ -0,0 +1,53 @@
+
+      @Override
+      public boolean isApplicable(@NotNull final Project project) {
+        return FileTypeIndex.containsFileOfType(HtmlFileType.INSTANCE, GlobalSearchScope.projectScope(project));
```
DartFileType

alexander-doroshko commented:
Thanks for the start! I think, we'll merge it later, when it's closer to a working state. I guess the main chunk of work is ahead. Daemon management and all the stuff. Also, I expect a conflict with JavaScriptDebugConfigurationProducer that we'll need to resolve somehow (maybe `RunConfigurationProducer.shouldReplace` will help).

jwren commented:
- Rebased on top of the latest master branch
- Updated with your comments

I'll try to keep this branch merge-able, and will keep the future commits not squashed into this initial commit so you can track the changes.

jwren commented:
@alexander-doroshko 

Some notes on the current status:

- The run configuration is still disabled in the `plugin.xml`,
- This run config worked locally today (Oct 4th) with webdev version `2.5.2-dev`, to use this version I modified the command line generated to run `webdev.dart` locally,
- Expression evaluation doesn't yet work, we should disable the action when doing a launch of this run config type,
- I modified the trying to connect message to not include the connect url as we don't know ahead of time what the connect url will be,
- The Windows issues, https://github.com/dart-lang/webdev/issues/709, will be fixed soon
- The Flutter plugin has removed the timeout when attempting to connect the debugger with the idea that users can cancel or wait as appropriate, we may want to consider the same change in the Dart plugin, or perhaps just for this launch configuration.
- The changes in `VmServiceWrapper.java` are conditions encountered with the Dart VM when debugging Flutter apps, which should harden our debugger implementation, I think there are some other changes I need to migrate

alexander-doroshko commented:
Thanks!
Two thoughts.

Regarding run configuration UI. As the whole process is based on the `webdev` tool, I think we should have options according to the [webdev](https://pub.dev/packages/webdev#webdev-serve) command line. As a start, directory name and port would probably be enough (that's what we effectively have in the current workflow). I think this would be better than HTML file path (which we, in fact, do not need).

We need to handle run/debug session termination better. Currently, when the user closes the Chrome window, Run/Debug session in the IDE remains visually active (and daemon remains alive). Do we want the daemon to stay alive between launches? It could make nest start faster though I'm not sure how much faster. For us, it's much easier to start/stop daemon as user starts/ends run/debug session, so we may start with this. But then we need to make sure that daemon stops on Chrome close.

Also, it would be great to see the webdev command line and 'Process finished with exit code x' in the console.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java:
```diff
@@ -116,6 +117,11 @@ public void received(final Isolate isolate) {
                       }
 
+                      // This is the entry point for attaching a debugger to a running app.
+                      if (eventKind == EventKind.Resume) {
```
Do you know how to reach this branch?

jwren replied:
I reverted the changes in this file.  When I saw the difference to what the Dart Plugin (as compared to the Flutter plugin) I originally thought that the changes were for webdev.  @stevemessick advice as necessary.

devoncarew commented:
> We need to handle run/debug session termination better. Currently, when the user closes the Chrome window, Run/Debug session in the IDE remains visually active (and daemon remains alive). Do we want the daemon to stay alive between launches? It could make nest start faster though I'm not sure how much faster.

My guess is that having one webdev instance per launch makes sense.

> For us, it's much easier to start/stop daemon as user starts/ends run/debug session, so we may start with this. But then we need to make sure that daemon stops on Chrome close.

webdev should be doing this - if it isn't, we should just file a bug on https://github.com/dart-lang/webdev.

alexander-doroshko commented:
> webdev should be doing this - if it isn't, we should just file a bug on https://github.com/dart-lang/webdev.

webdev server doesn't terminate on Chrome close

jwren commented:
@alexander-doroshko I just uploaded a few more commits on this PR, and just saw your comments, thanks for taking a look.  I'll follow up with the Webdev devs on the Dart team to see if the chrome tab closing issue is a known issue.  The Windows issues are currently a work in progress for them.

devoncarew commented:
> webdev server doesn't terminate on Chrome close (at least in my environment)

From talking to the webdev team, `webdev serve` won't terminate when the chrome process closes, but `webdev daemon` should.

jwren commented:
> Thanks!
> Two thoughts.
> 
> Regarding run configuration UI. As the whole process is based on the `webdev` tool, I think we should have options according to the [webdev](https://pub.dev/packages/webdev#webdev-serve) command line. As a start, directory name and port would probably be enough (that's what we effectively have in the current workflow). I think this would be better than HTML file path (which we, in fact, do not need).
> 

Having a `cwd` (current working directory) would be more in line with what the `webdev` command line would take, however it was a user experience conversation that lead me to include the `*.html` file instead.  When users go to launch some web application they think more in terms of some html file, not a directory.

As for including a port, the thought was that we could add it if users ask, but if possible, lets not include preferences were we don't need them.

@jakemac53 & @grouma -- would it be possible to have the `*.html` file be passed in as a parameter on the command line.  It seems reasonable that a user might have an `index.html`, but also some other `html` files that they might want to launch and debug.


> We need to handle run/debug session termination better. Currently, when the user closes the Chrome window, Run/Debug session in the IDE remains visually active (and daemon remains alive). Do we want the daemon to stay alive between launches? It could make the next start faster though I'm not sure how much faster. For us, it's much easier to start/stop daemon as user starts/ends run/debug session, so we may start with this. But then we need to make sure that daemon stops on Chrome close.
> 
> Also, it would be great to see the webdev command line and 'Process finished with exit code x' in the console.

@grouma  -- for context, this *is* run with `daemon` mode

alexander-doroshko commented:
Thanks! I'll merge it on Monday.

devoncarew commented on Dart/src/com/jetbrains/lang/dart/DartBundle.properties:
```diff
@@ -98,10 +98,18 @@ fields.to.generate.setters=Select Fields to Generate Setters
 remote.debug.configuration.description=Debug Remote Dart Command Line Application
+
+webdev.debug.configuration.name=Dart Web
+webdev.debug.configuration.description=Debug Dart Webdev
```
`Debug Dart Web App`?

jwren replied:
Same thought here as above.

jwren commented on Dart/src/com/jetbrains/lang/dart/DartBundle.properties:
```diff
@@ -98,10 +98,18 @@ fields.to.generate.setters=Select Fields to Generate Setters
 remote.debug.configuration.name=Dart Remote Debug
 remote.debug.configuration.description=Debug Remote Dart Command Line Application
+
+webdev.debug.configuration.name=Dart Web
```
It originally included 'App' (before my first cut of this PR), I am happy to include or exclude it.  Flutter has 'Flutter' not 'Flutter App'..., it felt redundant, but again, I can be convinced either way.

devoncarew replied:
Gotcha. Looking at other run configurations, it looks like they mostly omit any trailing 'App' text.

jwren commented on Dart/src/com/jetbrains/lang/dart/DartBundle.properties:
```diff
@@ -98,10 +98,18 @@ fields.to.generate.setters=Select Fields to Generate Setters
+webdev.debug.configuration.name=Dart Web
+webdev.debug.configuration.description=Debug Dart Webdev
+choose.html.main.file=Choose HTML File
+old.dart.sdk.for.webdev=Dart SDK {0}+ is required for Dart webdev debugging, current version: {1}
```
This message appears in the run config parameter error/ exception section which has limited space.  I'll iterate.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public DartWebdevConfigurationType() {
+    super("DartRemoteDebugConfigurationType",
```
copy/paste typo :)

alexander-doroshko commented:
Sorry, I failed to merge it today but will definitely merge tomorrow and we'll iterate further.

jwren commented on Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConsoleView.java:
```diff
@@ -0,0 +1,55 @@
+      super.print(text, contentType);
+    }
+    final String logMessage = DartDaemonParserUtil.getLogMessage(text.trim());
+    if (logMessage != null && !logMessage.isEmpty()) {
```
It is part of the log coming out of `webdev'.

CC @grouma

jwren commented:
Thanks for the comments @devoncarew.

alexander-doroshko commented:
Merged with some minor tweaks.
